### PR TITLE
useVelocity

### PR DESCRIPTION
**Pull request by mattgperry**
This PR adds a `useVelocity` hook. It accepts a motion value and returns a new one that updates with its velocity.

```javascript
const x = useMotionValue(0)
const xVelocity = useVelocity(x)
const scale = useTransform(xVelocity, [0, 1000], [1, 2], { clamp: false })
```

Fixes https://github.com/framer/motion/issues/1003

Demo: https://codesandbox.io/s/framer-motion-310-usevelocity-nw1dn?file=/src/Example.tsx

### CODE REVIEW

File: src/value/use-velocity.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export function useVelocity(value: MotionValue<number>): MotionValue<number> {
+    const velocity = useMotionValue(value.getVelocity())
+
+    useEffect(() => {
+        return value.velocityUpdateSubscribers.add((newVelocity) => {
+            velocity.set(newVelocity)
+        })
+    }, [value])
+
+    return velocity
```
Comment (nvh) [inline src/value/use-velocity.ts:24]: Nice that this is so simple 💪🏻

File: src/value/use-on-change.ts [hunk 2]
```diff
@@ -6,9 +6,9 @@ export function useOnChange<T>(
     value: MotionValue<T> | number | string,
     callback: Subscriber<T>
 ) {
-    useEffect(() =>
-        isMotionValue(value) ? value.onChange(callback) : undefined
-    )
+    useEffect(() => {
+        if (isMotionValue(value)) return value.onChange(callback)
+    }, [callback])
```
Comment (nvh) [inline src/value/use-on-change.ts:11]: Is this a driveby fix?

File: src/value/index.ts [hunk 3]
```diff
@@ -61,6 +61,13 @@ export class MotionValue<V = any> {
      */
     private updateSubscribers = subscriptionManager<Subscriber<V>>()
 
+    /**
+     * Functions to notify when the velocity updates.
+     *
+     * @internal
+     */
+    public velocityUpdateSubscribers = subscriptionManager<Subscriber<number>>()
```
Comment (nvh) [inline src/value/index.ts:69]: Why do we need a separate set of velocitySubscribers? Is that an optimisation because the velocity does not change as much?

File: src/value/use-velocity.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * ```
+ *
+ * @public
+ */
+export function useVelocity(value: MotionValue<number>): MotionValue<number> {
+    const velocity = useMotionValue(value.getVelocity())
+
+    useEffect(() => {
+        return value.velocityUpdateSubscribers.add((newVelocity) => {
+            velocity.set(newVelocity)
```
Comment (nvh) [inline src/value/use-velocity.ts:20]: We could theoretically just subscribe to the value and then call `getVelocity` on that, right?

**Review by nvh — APPROVED** [review]

**Comment by djfarly** [comment]
Tyvm @mattgperry – this api solves my use case perfectly and is much smarter that what I came up with. You work fast. 😬

File: src/value/use-velocity.ts (hunk 4 above)
Comment (mattgperry) [reply src/value/use-velocity.ts:20]: That was my preferred solution but there is a caveat that led me to putting it inside `MotionValue`. Every time a `MotionValue` is updated it schedules a velocity check at the end of the subsequent frame, so if it isn't updated in the meantime we will correctly report the velocity as `0` if requested.

It's this "if requested"  bit that's the problem. We'd have to duplicate that kind of scheduling here if we were to make sure the velocity is updated correctly.  By making explicit velocity subscribers we can ensure they get called at the same time velocity is updated, no matter the exact mechanism we use to perform that now or in the future.

File: src/value/use-on-change.ts (hunk 2 above)
Comment (mattgperry) [reply src/value/use-on-change.ts:11]: Sort of, it didn't look right to me but I also haven't heard of any breakages that I'm fixing.

File: src/value/use-velocity.ts (hunk 4 above)
Comment (nvh) [reply src/value/use-velocity.ts:20]: Makes sense 👍🏻

File: src/value/index.ts (hunk 3 above)
Comment (mattgperry) [reply src/value/index.ts:69]: Hopefully explained in the other comment
